### DISCUSSION TITLE
🍦 VanFS: 1:1 bindings from F# to 🍦VanJS (A tiny Reactive UI Framework without React/JSX) , based on Fable

### DISCUSSION THREAD

**Discussion by ken-okabe** [fable-compiler/Fable#3814] [Show and tell]
Based on **Fable**, I have developed: 
**🍦 VanFS**: 1:1 bindings from F# to 
  [**🍦 VanJS**](https://vanjs.org/) (an ultra-lightweight , zero-dependency , and unopinionated Reactive UI framework based on pure vanilla JavaScript and DOM without React/JSX) **+ WebComponents + FRP**

https://github.com/ken-okabe/vanfs

![image](https://github.com/fable-compiler/Fable/assets/1316994/8fef5d65-d411-4b86-8492-364fc1016247)

I appreciate the dedication and effort of all the Fable contributors, which have made my work possible.

I'd be grateful if you could take a look at my project whenever you have a chance. I'd really appreciate your feedback 😊

---
 _[**VanJS**](https://vanjs.org/) is a rising star among front-end frameworks._

https://risingstars.js.org/2023/en#section-framework

![image](https://github.com/fable-compiler/Fable/assets/1316994/23923f5b-5b9c-41b8-b57c-2db04b6cf02a)

**Comment by MangelMaxime** [comment]
Hello looks like there is a lot of content in your project great work.

You say that VanFS is a 1:1 binding however, I am seeing TypeScript code in the repository. Am I missing something or is the project more like a template / showcase of how VanJS can be consumed/loaded in a mixed solution?

**Comment by ken-okabe** [comment]
@MangelMaxime 
Hi, thanks for your feedback and sorry for the confusion, so probably, I should make things clearer in the document.

As https://github.com/ken-okabe/vanfs?tab=readme-ov-file#vanfs

**VanFS is a project template** for 1:1 bindings from [F#](https://fsharp.org/) to [VanJS](https://vanjs.org/), 
and the reason to use TypeScript is:

1. The internal mechanism of VanFS
2. For style and Web Components

# 1. The internal mechanism of VanFS

https://github.com/ken-okabe/vanfs/blob/main/van-api/ts/basic.ts

```ts
import van from 'vanjs-core'

import '../../web-imports/components'
import { cssURLs } from '../../web-imports/css-urls';

// css imports
cssURLs.forEach(url => {
    let stylesheet = document.createElement("link");
    stylesheet.rel = "stylesheet";
    stylesheet.href = url;
    document.head.appendChild(stylesheet);
});


// unary function ([a,b,c,...]) in F#
// -> n-ary function (a,b,c,...) in VanJS
let n =
    f => array =>
        f(...array);

export let tags =
    new Proxy(van.tags, {
        get: (target, property) => {
            return n(target[String(property)]);
        }
    });

export let add = n(van.add);
```

CSS imports 
+
The conversion for
// unary function ([a,b,c,...]) in F#
// -> n-ary function (a,b,c,...) in VanJS

**This is under the `van-api` directory which is essential and we would not want to modify it to keep things working.**

---

# 2. For style and Web Components

https://github.com/ken-okabe/vanfs/tree/main?tab=readme-ov-file#getting-started


![image](https://github.com/fable-compiler/Fable/assets/1316994/050881c6-43b8-40e0-a6c9-ef1341d581b3)

---

https://github.com/ken-okabe/vanfs/tree/main?tab=readme-ov-file#web-components

# 🌐 Web Components

## Install the Fluent UI Web Components with NPM or Alternatives

https://www.npmjs.com/package/@fluentui/web-components

```sh
npm install @fluentui/web-components
```

## Import and Register the web components

Let's use Fruent  Web **Card**  and  **Checkbox**.

https://learn.microsoft.com/en-us/fluent-ui/web-components/components/card?pivots=typescript

https://learn.microsoft.com/en-us/fluent-ui/web-components/components/checkbox?pivots=typescript

![image](https://raw.githubusercontent.com/ken-okabe/web-images4/main/img_1711754186205.png)

#### /web-imports/components.ts

<img width="100%" src="https://raw.githubusercontent.com/ken-okabe/web-images/main/typescript.svg">

```ts
import {
    provideFluentDesignSystem,
    fluentCard,
    fluentCheckbox
} from "@fluentui/web-components";

provideFluentDesignSystem()
    .register(
        fluentCard()
    );

provideFluentDesignSystem()
    .register(
        fluentCheckbox()
    );
```

---

Like https://learn.microsoft.com/en-us/fluent-ui/web-components/components/checkbox?pivots=typescript

![image](https://github.com/fable-compiler/Fable/assets/1316994/ccbaa8a8-77bc-407d-82d3-f398a3b8b642)

Using Web Components requires adding some TypeScript code, and we simply can copy-paste the code to

![image](https://github.com/fable-compiler/Fable/assets/1316994/d052a5aa-df77-4819-9e0e-b8563a68076a)

or

---


## Material Web Components use Google Fonts/Icons

https://m3.material.io/styles/icons/overview


## Install Google Fonts/Icons

Obtain required CSS URLs from the Google Fonts page.

https://fonts.google.com/icons?selected=Material+Symbols+Outlined:star:FILL@0;wght@400;GRAD@0;opsz@24

![image](https://raw.githubusercontent.com/ken-okabe/web-images4/main/img_1712042147933.png)

Add the CSS URL to

![image](https://raw.githubusercontent.com/ken-okabe/web-images4/main/img_1711754186205.png)

#### /web-imports/css-urls.ts

<img width="100%" src="https://raw.githubusercontent.com/ken-okabe/web-images/main/typescript.svg">

```ts
export let cssURLs = [
 "https://fonts.googleapis.com/css2?family=Material+Symbols+Outlined:opsz,wght,FILL,GRAD@20..48,100..700,0..1,-50..200"
];
```
---

---

The point is **VanJS does not handle the installation of CSS or Web Components.**, so
**🍦 VanFS**: 1:1 bindings from F# to 
  [**🍦 VanJS**](https://vanjs.org/) (an ultra-lightweight , zero-dependency , and unopinionated Reactive UI framework based on pure vanilla JavaScript and DOM without React/JSX) **+ WebComponents + FRP**

Again, I probably should explain clearer in the document.

Feel free to provide any further questions or suggestions you might have please. Thanks.
